### PR TITLE
feat(performance): #2539 Add perf telemetry + infrastructure

### PR DESCRIPTION
**Pull request by dmose**
This adds basic performance telemetry infrastructure and the first bits of telemetry that use it for the new tab.  I suspect it'll mostly be easier to review the individual commits.

How to test if it's working

1. Edit [system-addon/lib/Activity-Stream.jsm](system-addon/lib/Activity-Stream.jsm) and 
set the `telemetry.log` preference value to `true`.
2. `npm run startmc`
3. `cd ../mozilla-central && ./mach build && ./mach run`
4. Open the Browser Console
5. Create a new tab
6. Close it
7. Make sure the logged session ping has an appropriate .perf object

### CODE REVIEW

**Comment by dmose** [comment]
Note that there will be a few spin-off bugs here; one on handling a few edge cases, and one for removing the PerfPing infrastructure entirely.

File: system-addon/test/schemas/pings.js [hunk 1]
```diff
@@ -63,14 +63,34 @@ const SessionPing = Joi.object().keys(Object.assign({}, baseKeys, {
...
   session_duration: Joi.number().integer().required(),
-  action: Joi.valid("activity_stream_session").required()
+  action: Joi.valid("activity_stream_session").required(),
+  perf: Joi.object().keys({
+    // Timestamp of the action perceived by the user to trigger the load
+    // of this page.
+    //
+    // Not required at least for error cases where the
+    // observer event doesn't fire
+    load_trigger_ts: Joi.number().positive(),
```
Comment (dmose) [inline system-addon/test/schemas/pings.js]: None of the numbers are technically required, though in some ideal world they'd be there every time.  I just suspect that there's going to be some complex code paths where one or another event doesn't fire, and I'd like us to know about them.  That said, I guess we could mark them as required and just see if anything goes wrong.  Thoughts?

File: system-addon/test/schemas/pings.js [hunk 2]
```diff
@@ -63,14 +63,34 @@ const SessionPing = Joi.object().keys(Object.assign({}, baseKeys, {
...
+    // doesn't fire
+    load_trigger_type: Joi.valid(["menu_plus_or_keyboard"]),
+    // When the page itself receives an event that document.visibilityState
+    // == visible.
+    //
+    // Not required at least for the (error?) case where the
+    // visibility_event doesn't fire.  (It's not clear whether this
+    // can happen in practice, but if it does, we'd like to know about it).
+    visibility_event_rcvd_ts: Joi.number().positive()
+  }).required()
```
Comment (dmose) [inline system-addon/test/schemas/pings.js]: I suppose it's not technically required, if none of the child properties are there.  So the answer sort of depends on how we decide the other thing...

File: system-addon/common/PerfService.jsm [hunk 3]
```diff
@@ -0,0 +1,94 @@
+/* globals Services */
+"use strict";
+
+let usablePerfObj;
+
+let Cu;
+
+/* istanbul ignore if */
+if (typeof Window === "undefined") {
```
Comment (dmose) [inline system-addon/common/PerfService.jsm]: Good idea; will fix.

File: system-addon/lib/TelemetryFeed.jsm [hunk 4]
```diff
@@ -24,23 +27,47 @@ this.TelemetryFeed = class TelemetryFeed {
...
    * @param  {string} id the portID of the open session
+   * @param  {number} absVisChangeTime absolute timestamp of
+   *                                   document.visibilityState becoming visible
    */
-  addSession(id) {
+  addSession(id, absVisChangeTime) {
+    // XXX note that there is a race condition here; we're assuming that the
+    // most recent tab
+    let absBrowserOpenTabStart =
+      perfService.getMostRecentAbsMarkStartByName("browser-open-newtab-start");
```
Comment (dmose) [inline system-addon/lib/TelemetryFeed.jsm]: It's already a public API point, I'm not convinced it really buys us much.  Search-and-replace for something like this is awfully easy within a file, after all.

File: system-addon/test/unit/common/PerfService.test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        // called.
+        assert.equal(absMarkStart - perfService.timeOrigin, 2);
+      });
+  });
+
+  describe("#mark", () => {
+    it("should call the wrapped version of mark", () => {
+      sandbox.spy(fakePerfObj, "mark");
+
+      perfService.mark("monkey");
```
Comment (dmose) [inline system-addon/test/unit/common/PerfService.test.js]: 🙉

File: system-addon/test/schemas/pings.js (hunk 1 above)
Comment (dmose) [inline system-addon/test/schemas/pings.js]: Leaving the .required() off after discussion.  I've updated the schema in pings.js and the data dictionary with indicators that we want this tracked on the server side.

**Comment by dmose** [comment]
@ncloudioj, can you review the changes to data_events.md, data_dictionary.md, and pings.js please?

**Comment by dmose** [comment]
@fmarier, would you please review the changes to data_events.md, data_dictionary.md from a privacy standpoint?

**Comment by fmarier** [comment]
datareview+

**Comment by dmose** [comment]
@k88hudson can you take a look at the last three commits?  All they do is change the telemetry pref that we're looking at to `datareporting.healthreport.uploadEnabled`...

Thanks!

**Comment by as-pine-proxy** [comment]
[Treeherder: c63f9f5eb3cfdbfc6e8be03a87f8eb25fe3a0a11](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=c63f9f5eb3cfdbfc6e8be03a87f8eb25fe3a0a11)
